File: core/src/users/permissions.rs
//! users 模块权限常量。
//!
//! 本模块属于 `users` 业务层，负责定义用户域需要声明给安全层和 RBAC 启动逻辑的稳定权限代码。
//! 它不负责权限定义补齐或权限分配存储。

/// 允许创建新用户的权限代码。
pub(crate) const REGISTER_USER_PERMISSION: &str = "user.register";

/// 查看用户列表和用户详情的权限代码。
pub(crate) const READ_USER_PERMISSION: &str = "user.read";

/// 启用或停用用户账号的权限代码。
pub(crate) const UPDATE_USER_STATUS_PERMISSION: &str = "user.status.update";

/// 整体替换用户权限的权限代码。
pub(crate) const UPDATE_USER_PERMISSIONS_PERMISSION: &str = "user.permissions.update";

/// 查看权限定义的权限代码。
pub(crate) const READ_USER_PERMISSION_DEFINITION_PERMISSION: &str = "user.permission.read";

/// 管理员直接重置其他用户密码的权限代码。
pub(crate) const RESET_USER_PASSWORD_PERMISSION: &str = "user.password.reset";

File: frontend/src/auth/permissions.ts
// 本文件定义 frontend 使用的稳定权限代码和只读判断；它不替代后端实时授权检查。

/** 用户管理页面和操作使用的权限代码。 */
export const userPermissions = {
  register: "user.register",
  read: "user.read",
  updateStatus: "user.status.update",
  delete: "user.delete",
  updatePermissions: "user.permissions.update",
  readPermissionDefinitions: "user.permission.read",
  resetPassword: "user.password.reset",
  updateUsername: "user.username.update",
} as const;

/** 库存业务页面使用的稳定权限代码。 */
export const stockPermissions = {
  dashboardRead: "stock.dashboard.read",
  itemRead: "stock.item.read",
  itemManage: "stock.item.manage",
  locationRead: "stock.location.read",
  locationManage: "stock.location.manage",
  inboundCreate: "stock.inbound.create",
  inboundRead: "stock.inbound.read",
  inboundApprove: "stock.inbound.approve",
  outboundCreate: "stock.outbound.create",
  outboundRead: "stock.outbound.read",
  outboundApprove: "stock.outbound.approve",
  templateRead: "stock.template.read",
  templateManage: "stock.template.manage",
  substituteRead: "stock.substitute.read",
  substituteManage: "stock.substitute.manage",
  auditRead: "audit.read",
} as const;

/** 判断权限快照是否包含指定权限。 */
export function hasPermission(
  permissions: readonly string[] | undefined,
  requiredPermission: string | undefined,
): boolean {
  return requiredPermission === undefined || permissions?.includes(requiredPermission) === true;
}

/** 判断权限快照是否同时包含页面或操作要求的全部权限。 */
export function hasPermissions(
  permissions: readonly string[] | undefined,
  requiredPermissions: readonly string[] | undefined,
): boolean {
  return (
    requiredPermissions === undefined ||
    requiredPermissions.every((permission) => permissions?.includes(permission) === true)
  );
}
